Span: add attribute, event, link setter to the API

alban opened the pull request:
Basic tests that exercise the API are added as well. However it does not
actually test that the attributes, events and links are written properly
because there is no getter. Those tests can be added later when the
attributes are actually used.

The spec mentions additional APIs for lazy initialization (AddLazyEvent,
AddLazyLink). Those are not added in this patch because OT-Python does
not expose the Event type (currently defined in the SDK only).

---

https://github.com/open-telemetry/opentelemetry-python/issues/75 (attribute)
https://github.com/open-telemetry/opentelemetry-python/issues/76 (event)
https://github.com/open-telemetry/opentelemetry-python/issues/77 (link)

a-feld commented on opentelemetry-api/src/opentelemetry/trace/__init__.py:
```diff
@@ -65,6 +65,8 @@
 import typing
 
 from opentelemetry import loader
+from opentelemetry import trace as trace_api
```
This import can be removed and the `trace_api.SpanContext` below can be replaced with `'SpanContext'`! 😄

alban replied:
Thanks! I'm fixing this...

carlosalberto commented:
It all looks good, in addition to what was commented by @a-feld mentioned

Also, the tickets were originally created to track the API + the SDK changes, so that's remaining ;) (Can be done in future PRs)

reyang commented on opentelemetry-sdk/tests/trace/test_trace.py:
```diff
@@ -114,6 +114,34 @@ def test_start_span_explicit(self):
+
+            root.add_event('event1', 'event1-detail')
+
+            root.add_link(other_neighbor, 'neighbor')
```
`add_link` requires a `SpanContext`, do we intend to revisit this later?

alban replied:
I am not sure to follow: on this line, the `other_neighbor` argument is a `SpanContext`, created above:
```
        other_neighbor = trace_api.SpanContext(
            trace_id=0x0000000000000000000000000000cafe,
            span_id=0x00000000cafe1234
        )
```

reyang replied:
Sorry I thought it was a string. You're right.

I guess the correct invocation should be something like this? `root.add_link(other_neighbor, {'name': 'neighbor'})`
I also wonder if the API should treat `attributes` as optional? `root.add_link(my_span_context)`

carlosalberto replied:
> I also wonder if the API should treat attributes as optional? 

Oh actually it should (same with `add_event()`)

Related question: should we add `Link`/`Event` classes or keep with this minimalistic approach? cc @c24t

c24t replied:
This PR seems like a great place to add `Link`/`Event`.

c24t replied:
@reyang points out that we could use a singleton default empty attribute map here (like  @carlosalberto's [`CURRENT_SPAN`](https://github.com/c24t/opentelemetry-python/blob/e8ef980daf3c72abc699a654dfe6c4e5b8bfbcda/opentelemetry-api/src/opentelemetry/trace/__init__.py#L232) to avoid creating a bunch of empty attributes.

toumorokoshi commented on opentelemetry-sdk/tests/trace/test_trace.py:
```diff
@@ -114,6 +114,34 @@ def test_start_span_explicit(self):
+        tracer = trace.Tracer(context)
+
+        other_neighbor = trace_api.SpanContext(
+            trace_id=0x0000000000000000000000000000cafe,
```
thoughts on using the sdk's trace.get_trace_id() and trace.get_span_id()?

alban replied:
I'll use `trace.generate_trace_id()` and `trace.generate_span_id()`.

alban commented:
> Also, the tickets were originally created to track the API + the SDK changes, so that's remaining ;) (Can be done in future PRs)

Should I remove the "Fixes #XXX" in the PR description then?

carlosalberto commented:
@alban Yes, please ;) A simple mention to them should be fine (we can close later on the issues when the actual SDK portions are done)

c24t commented on opentelemetry-api/src/opentelemetry/trace/__init__.py:
```diff
@@ -102,6 +103,34 @@ def get_context(self) -> 'SpanContext':
+                  name: str,
+                  attributes: 'types.Attributes',
+                  ) -> None:
+        """Adds an Event.
```
One open question is whether the caller ever needs access to the events or links they create (in which case we need the classes in the API) or if we should treat them as part of the implementation of these methods.

carlosalberto replied:
There are two places where I've seen `Event`/`Link` and related members being publicly used/exposed : 1) `SpanData` (but as that is probable going to the SDK at the very least) and 2) `MessageEvent` (which extends `Event`, at least in Java: https://github.com/open-telemetry/opentelemetry-java/blob/master/contrib/trace_utils/src/main/java/io/opentelemetry/contrib/trace/MessageEvent.java).

Makes me *feel* we should include these `Event`, etc classes here.

c24t commented on opentelemetry-api/src/opentelemetry/trace/__init__.py:
```diff
@@ -102,6 +103,34 @@ def get_context(self) -> 'SpanContext':
+                 context: 'SpanContext',
+                 attributes: 'types.Attributes',
+                 ) -> None:
+        """Adds a Link to another Span.
```
```suggestion
        """Adds a Link to another span.
```

It's a pedantic change to be sure but the link here is to another conceptual span, not a `Span` in memory.

c24t commented on opentelemetry-api/src/opentelemetry/trace/__init__.py:
```diff
@@ -102,6 +103,34 @@ def get_context(self) -> 'SpanContext':
+                      ) -> None:
+        """Sets an Attribute.
+
+        Sets a single Attribute with the key and value passed as arguments.
```
```suggestion
        Sets a single `Attribute` with the key and value passed as arguments.
```

alban replied:
This makes the CI complain:
```
/home/travis/build/open-telemetry/opentelemetry-python/opentelemetry-api/src/opentelemetry/trace/__init__.py:docstring of opentelemetry.trace.Span.set_attribute:3: WARNING: 'any' reference target not found: Attribute
/home/travis/build/open-telemetry/opentelemetry-python/opentelemetry-api/src/opentelemetry/trace/__init__.py:docstring of opentelemetry.trace.Span.add_link:3: WARNING: 'any' reference target not found: Link
```

reyang replied:
It needs to be a full qualified name.

c24t commented on opentelemetry-api/src/opentelemetry/trace/__init__.py:
```diff
@@ -102,6 +103,34 @@ def get_context(self) -> 'SpanContext':
+                  ) -> None:
+        """Adds an Event.
+
+        Adds a single Event with the name and, optionally, attributes passed
```
```suggestion
        Adds a single `Event` with the name and, optionally, attributes passed
```

alban replied:
At the moment, we don't. We have the following
```
Event = namedtuple('Event', ('name', 'attributes'))
```
But that's in the SDK, not exposed in the API.

c24t commented on opentelemetry-api/src/opentelemetry/trace/__init__.py:
```diff
@@ -102,6 +103,34 @@ def get_context(self) -> 'SpanContext':
+                 ) -> None:
+        """Adds a Link to another Span.
+
+        Adds a single Link to the Span via the SpanContext passed as argument.
```
```suggestion
        Adds a single `Link` to the `Span` via the `SpanContext` passed as argument.
```

alban replied:
"the Span" should be "the span" (following your rationale https://github.com/open-telemetry/opentelemetry-python/pull/83#discussion_r313568783)

alban commented:
I updated the PR:

- tests: use a dict in add_link and add_event instead of a string by mistake
- make argument optional in add_link and add_event. I reused `Span.empty_attributes` in the implementation but didn't expose `Link`/`Event` in the API
- API docstring: changes in spelling "span" following Chris' suggestion. But I didn't take `Attribute` and `Link` to avoid CI error
- tests: we don't have getters but I added tests accessing the span members directly

Oberon00 commented on opentelemetry-api/src/opentelemetry/trace/__init__.py:
```diff
@@ -102,6 +103,35 @@ def get_context(self) -> 'SpanContext':
+                 ) -> None:
+        """Adds a Link to another span.
+
+        Adds a single Link to the span via the `SpanContext` passed as
```
```suggestion
        Adds a single Link from this Span to another Span identified by the `SpanContext` passed as
```

Oberon00 commented on opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py:
```diff
@@ -238,18 +238,22 @@ def set_attribute(self: 'Span',
         if self.events is Span.empty_events:
             self.events = BoundedList(MAX_NUM_EVENTS)
+        if attributes is None:
+            attributes = Span.empty_attributes
```
It somewhat irks me that we can't use `Span.empty_attributes` as default and leave out this check, because the API already specifies a default argument of `None`. But I know of no better way to write this than the one you used here.

alban replied:
I initially tried to use `{}` as default value but the CI complained with:
```
Dangerous default value {} as argument
```
and I read recommendations to use `None` in that case.

alban commented on opentelemetry-api/src/opentelemetry/trace/__init__.py:
```diff
@@ -102,6 +103,35 @@ def get_context(self) -> 'SpanContext':
+
+    def add_event(self: 'Span',
+                  name: str,
+                  attributes: 'types.Attributes' = None,
```
Ok, I'll pick this.

alban commented:
Thanks for the additional review. I updated the PR again:
- use your phrasing in the docstring
- rename argument to `link_target_context`
- avoid types in quotes when possible (`types.Attributes`)
- update ASK to follow changes in API
